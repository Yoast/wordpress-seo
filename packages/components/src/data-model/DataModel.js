import React from "react";
import PropTypes from "prop-types";
import { __, sprintf } from "@wordpress/i18n";

import "./data-model.css";

/**
 * Props for dataItem.
 */
const dataItemProps = {
	width: PropTypes.number.isRequired,
	name: PropTypes.string.isRequired,
	number: PropTypes.number.isRequired,
};

/**
 * Creates a list item from the props.
 *
 * @param {object} props The properties.
 *
 * @returns {HTMLElement} A list item.
 */
const DataItem = ( props ) => {
	/* Translators: %d expands to number of occurrences. */
	const screenReaderText = sprintf( __( "%d occurrences", "yoast-components" ), props.number );
	return (
		<li
			key={ props.name + "_dataItem" }
			style={ { "--yoast-width": `${ props.width}%` } }
		>
			{ props.name }
			<span>{ props.number }</span>
			<span className="screen-reader-text">{ screenReaderText }</span>
		</li>
	);
};

DataItem.propTypes = dataItemProps;

/**
 * Creates a DataModel based on props.
 *
 * @param {object} props The properties.
 *
 * @returns {HTMLElement} A <ul> with <li> items.
 */
const DataModel = ( props ) => (
	<ul
		className="yoast-data-model"
		aria-label={ __( "Prominent words", "yoast-components" ) }
	>
		{ props.items.map( DataItem ) }
	</ul>
);

DataModel.propTypes = {
	items: PropTypes.arrayOf( PropTypes.shape( dataItemProps ) ),
};

DataModel.defaultProps = {
	items: [],
};

export default DataModel;
