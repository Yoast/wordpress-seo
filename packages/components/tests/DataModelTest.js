import React from "react";
import renderer from "react-test-renderer";

import DataModel from "../src/data-model/DataModel";

describe( "DataModel", () => {
	it( "should render with only the required props", () => {
		const component = renderer.create( <DataModel
			items={ [ {
				name: "test",
				width: 60,
				number: 60,
			} ] }
		/> );
		const tree = component.toJSON();
		expect( tree ).toMatchSnapshot();
	} );
} );
