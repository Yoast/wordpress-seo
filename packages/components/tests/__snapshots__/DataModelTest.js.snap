// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataModel should render with only the required props 1`] = `
<ul
  aria-label="Prominent words"
  className="yoast-data-model"
>
  <li
    style={
      Object {
        "--yoast-width": "60%",
      }
    }
  >
    test
    <span>
      60
    </span>
    <span
      className="screen-reader-text"
    >
      60 occurrences
    </span>
  </li>
</ul>
`;
